miden/src/lib.rs

```diff
@@ -1,84 +1,70 @@
+/// Executes the specified `program` and returns the result together with a STARK-based proof of
+/// the program's execution.
 ///
-/// * `inputs` specifies the initial stack state and provides secret input tapes;
-/// * `num_outputs` specifies the number of elements from the top of the stack to be returned;
+/// * `inputs` specifies the initial state of the stack as well as non-deterministic (secret)
+///   inputs for the VM.
+/// * `num_stack_outputs` specifies the number of elements from the top of the stack to be
+///   returned.
+/// * `options` defines parameters for STARK proof generation.
+///
+/// # Errors
+/// Returns an error if program execution or STARK proof generation fails for any reason.
 pub fn execute(
-    program: &Program,
+    program: &Script,
     inputs: &ProgramInputs,
-    num_outputs: usize,
+    num_stack_outputs: usize,
     options: &ProofOptions,
-) -> Result<(Vec<u128>, StarkProof), ProverError> {
-    assert!(
-        num_outputs <= MAX_OUTPUTS,
-        "cannot produce more than {} outputs, but requested {}",
-        MAX_OUTPUTS,
-        num_outputs
-    );
+) -> Result<(Vec<u64>, StarkProof), ExecutionError> {
+    if num_stack_outputs > MIN_STACK_DEPTH {
+        return Err(ExecutionError::TooManyStackOutputs(num_stack_outputs));
+    }
 
     // execute the program to create an execution trace
     #[cfg(feature = "std")]
     let now = Instant::now();
-    let trace = processor::execute(program, inputs);
+    let trace = processor::execute(program, inputs)?;
     #[cfg(feature = "std")]
     debug!(
-        "Generated execution trace of {} registers and {} steps in {} ms",
+        "Generated execution trace of {} columns and {} steps in {} ms",
         trace.width(),
         trace.length(),
         now.elapsed().as_millis()
     );
 
-    // copy the user stack state the the last step to return as output
-    let last_state = get_last_state(&trace);
-    let outputs = last_state.user_stack()[..num_outputs]
+    // copy the stack state at the last step to return as output
+    let outputs = trace.last_stack_state()[..num_stack_outputs]
         .iter()
         .map(|&v| v.as_int())
         .collect::<Vec<_>>();
 
-    // make sure number of executed operations was sufficient
-    assert!(
-        last_state.op_counter().as_int() as usize >= MIN_TRACE_LENGTH,
-        "a program must consist of at least {} operation, but only {} were executed",
-        MIN_TRACE_LENGTH,
-        last_state.op_counter()
-    );
-
-    // make sure program hash generated by the VM matches the hash of the program
-    let program_hash: [u8; 32] = last_state.program_hash().to_bytes().try_into().unwrap();
-    #[cfg(feature = "std")]
-    assert!(
-        *program.hash() == program_hash,
-        "expected program hash {} does not match trace hash {}",
-        hex::encode(program.hash()),
-        hex::encode(program_hash)
-    );
-
```

Comment:
Are these checks coming back later? If so, maybe a TODO placeholder would be helpful

Reply:
The first check is no longer need in the new structure. In fact, `clk` value at the last step would be equal to trace length minus 1.

For the second check, I'm thinking of moving it to the `execute()` method of the `processor`. I've added a TODO for it there.

Reply:
Makes sense, thanks!